issue-4-functions-in-dart/issue-4-functions-in-dart.md

```diff
@@ -68,62 +68,65 @@ minus the `mov eax 0xdeadbeef`, that's just the value to printed later, but the
+  uword stack_overflow_flags = thread->GetAndClearStackOverflowFlags();
+
+  // If an interrupt happens at the same time as a stack overflow, we
+  // process the stack overflow now and leave the interrupt for next
+  // time.
+  if (!thread->os_thread()->HasStackHeadroom() ||
+      IsCalleeFrameOf(thread->saved_stack_limit(), stack_pos)) {
+    if (FLAG_verbose_stack_overflow) {
+      OS::PrintErr("Stack overflow\n");
+      OS::PrintErr("  Native SP = %" Px ", stack limit = %" Px "\n", stack_pos,
+                   thread->saved_stack_limit());
+      OS::PrintErr("Call stack:\n");
+      OS::PrintErr("size | frame\n");
+      StackFrameIterator frames(ValidationPolicy::kDontValidateFrames, thread,
+                                StackFrameIterator::kNoCrossThreadIteration);
+      uword fp = stack_pos;
+      StackFrame* frame = frames.NextFrame();
+      while (frame != NULL) {
+        uword delta = (frame->fp() - fp);
+        fp = frame->fp();
+        OS::PrintErr("%4" Pd " %s\n", delta, frame->ToCString());
+        frame = frames.NextFrame();
       }
     }
-  }
 
-  if (first_stack_overflow_instr != NULL) {
-    first_stack_overflow_instr->RemoveFromGraph();
+    // Use the preallocated stack overflow exception to avoid calling
+    // into dart code.
+    const Instance& exception =
+        Instance::Handle(isolate->group()->object_store()->stack_overflow());
+    Exceptions::Throw(thread, exception);
+    UNREACHABLE();
   }
-}
+
+#if !defined(PRODUCT) && !defined(DART_PRECOMPILED_RUNTIME)
+  HandleStackOverflowTestCases(thread);
+#endif  // !defined(PRODUCT) && !defined(DART_PRECOMPILED_RUNTIME)
 ```
 
 so what all of this is doing is, with excerpts from Vyacheslav, this is checking the current stack pointer against *a limit* to make sure it's not gone over that, so that a runtime routine can catch such stack overflows and handle that internally! so basically we don't have to worry about that part of the code, is what I'm trying to say 🤠
 
-so until this point, before the `loc_9a63f` label, the 64-bit CPU register `eax` (the lower 32-bits of `rax`) is holding the value we are going to print to the console using the `print` function.
+so until this point, before the `loc_9a63f` label, the 32-bit CPU register `eax` (the lower 32-bits of `rax`) is holding the value we are going to print to the console using the `print` function. It's good to know, for me as well, that while operating in a 64-bit environment, a `mov` instruction onto a 32-bit register zeroes out the upper 32 bits of its 64-bit register, so that `mov eax, foo` will zero out the upper 32-bits of `rax`, `mov ebx, foo` will zero out the upper 32-bits of `rbx` and etc. I tried to find information about this in *Intel® 64 and IA-32 Architectures Software Developer’s Manual Volume 2 (2A, 2B, 2C & 2D): Instruction Set Reference, A-Z with no luck*. If you have a reference to this, I would love to have a tip please so we can add it to the bottom of this issue.
```

Comment:
It's in the section 3.4.1.1 in the Volume 1, which contains the following text (highlighted the relevant bullet point): 

> When in 64-bit mode, operand size determines the number of valid bits in the destination general-purpose register:
> • 64-bit operands generate a 64-bit result in the destination general-purpose register.
> • __32-bit operands generate a 32-bit result, zero-extended to a 64-bit result in the destination general-purpose register.__
> • 8-bit and 16-bit operands generate an 8-bit or 16-bit result. The upper 56 bits or 48 bits (respectively) of the destination general-purpose register are not modified by the operation. If the result of an 8-bit or 16-bit operation is intended for 64-bit address calculation, explicitly sign-extend the register to the full 64-bits.